RV-Monitor/Kernel_Safety_Claims_by_Runtime_Verification_Monitors.md

```diff
@@ -0,0 +1,542 @@
+* **FM3: **RV Monitor code is corrupted due to random interference from Kernel code. \
+**Effect of Failure: **RV Monitor code could not execute or wrongly execute leading to the monitor being ineffective. \
+**Prevention or detection measures: **RV Monitor code is stored in memory mapped with the RO attribute, hence any spurious write would trigger an SMMU exception that would, at least, halt the execution of the Kernel, hence resulting in the external watchdog to trigger and drive the platform safe state
+
+
+# RV Monitors Design
+
+
+### &lt;from failure modes to automata finite state diagrams>
+
+
+###  
+
+
+### https://docs.kernel.org/trace/rv/deterministic_automata.html
+
+
+### &lt;from automata to RV monitor implementation>
+
+
+# RV Monitors Examples
+
+
+## i6300esb Watchdog timeout monitor
+
+
+### Introduction and Requirements
+
+The goal of this RV Monitor is to qualify the Kernel according to a subset of the safety requirements defined for the Telltale Safety Application in this [sheet](https://docs.google.com/spreadsheets/d/1EbuVvhXo-xZc2aPTfMgQtPNPDQYtcozs/edit#gid=584539121). More specifically:
+
+
+
+* **KSR_0003**: the watchdog subsystem shall ensure the opened WD device to be the one specified as input argument
+* **KSR_0004**: The watchdog subsystem shall ensure the WD timeout to be set according to the IOCTL input parameter
+* **KSR_0005**: The operating system shall ensure the WD timeout to be not wrongly re-set to a different timeout value
+* **KSR_0012**: Writing the Watchdog shall reset it with a timeout equal to the one specified in the IOCTL. No other timeout value shall be used / reconfigured
+
+**Assumption:** it is assumed that any boot time operation has been successfully completed; we could consider the design of a RV Monitor for guarding against boot time misconfigurations as a follow-on activity
+
+
+### Static and Dynamic Architecture
+
+![alt_text](./i6300esb_watchdog_timeout.png "image_tooltip")
+ \
+As a first step we consider only the architectural elements that are functionally supporting the implementation of the safety requirements mentioned above; interfering architectural elements will be added later in the picture. \
+ \
+In the use case under analysis, to start with, we can identify 3 main elements interacting with each other:
+
+
+
+* The Safety Application
+* The Linux Kernel
+* The HW SoC that also includes a programmable i6300esb watchdog.
+
+ \
+Accordingly we have the following events that characterize the expected interactions in absence of failures.
+
+
+
+1. The Safety Application invokes the open syscall passing the input path from the filesystem associated with the watchdog to be opened.
```

Comment:
This part is somewhat inconsistent: The kernel is not per-se safe, nor is the application, therefore they can fail before configuring and activating the WD.

We could say that there needs to be a further external safety element which is qualified to detect this event.
Or we could say that the WD is wired (some can be configured so) to start in a predefined mode, at power on.
Either way, all that matters is the fact that this one must be either properly correctly configured by a safe element, or it must be self-configuring.
But safey bootstrapping of the WD is not a valid proposition.

Reply:
the application is assumed to be ASIL B, hence safe...

Reply:
The application can be implemented to be ASILB, however that doesn't preclude that the running application has been corrupted somehow, during loading, and therefore what is effectively being executed is not safe at all.

Reply:
Somewhere we can say that
a) We assume the safety app to be qualified at least up to the same ASIL that we target for the monitored Kernel code
b) We assume the Kernel code supporting the creation of the safety app address space and maintaining the integrity of the same to be already qualified at least up to the same ASIL as the safety app

Would that work?

Reply:
It's an impossible-to-prove AoU :-D 
How can that work?

Reply:
So WRT to a) it is a fair assumption since it is out of the scope of the operating system. Now WRT to b) we can say
"b) We assume the Kernel code supporting the creation of the safety app address space and maintaining the integrity of the same to be already qualified at least up to the same ASIL as the safety app. Such claims are in the scope of ELISA activities but not covered as part of this document"

Reply:
a) agreed
b) "If my grandmother had a wheel, she would have been a wheelbarrow" (Apologies, I coulnd't help it :-D )
In general I would not disagree, but here the assumption seems to overlap the scope of the document, which is about integrity (violations) in kenrel space.

Reply:
I was going to reply with a sentence about grandads and flippers but then I changed my mind... :)

This doc is explaining how we qualify the Kernel against the claims allocated to it by the safety app. Now you're saying "Hey look, we haven't proved that the Kernel is qualified against the claims of creating the app address space and maintaining its integrity", that is true, however it is not in the scope of the claims that I am analysing (hence the AoUs of the Kernel being qualified, somehow, against such claims).
The idea is that at some stage (I hope sooner, rather than later) we can also analyze what would take to qualify the Kernel against such claims, that are indeed also way more interesting than the one we are looking at.

So my ask is "please let's write this AoU for now and later we can address this"; actually I can add a note explaining that such a claim has to be considered as part of the future ELISA activities...

Reply:
Following the last WG discussion I will add a clear disclaimer at the top of the doc as in:

Disclaimer
This document only focuses on the qualification of the Kernel functionalities supporting the following requirements:
KSR_0003: the watchdog subsystem shall ensure the opened WD device to be the one specified as input argument;
KSR_0004: The watchdog subsystem shall ensure the WD timeout to be set according to the ioctl input parameter;
KSR_0005: The operating system shall ensure the WD timeout to be not wrongly re-set to a different timeout value;
KSR_0012: Writing the Watchdog shall reset it with a timeout equal to the one specified in the IOCTL. No other timeout value shall be used / reconfigured
All other Kernel Safety Requirements defined in [Telltale_Safe](https://docs.google.com/spreadsheets/d/1EbuVvhXo-xZc2aPTfMgQtPNPDQYtcozs/edit#gid=584539121) are not in the scope of this document.
Therefore the solution proposed in this document is NOT sufficient to qualify the Kernel against all the allocated safety requirements; only a subset of them is covered.

Reply:
ok